core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -0,0 +1,529 @@
+                for (TaskDescriptor task : tasks) {
+                    queuedPartitions.add(task.getPartitionId());
+                    partitionToTaskDescriptorMap.put(task.getPartitionId(), task);
+                    sinkExchange.ifPresent(exchange -> {
+                        ExchangeSinkHandle exchangeSinkHandle = exchange.addSink(task.getPartitionId());
+                        partitionToExchangeSinkHandleMap.put(task.getPartitionId(), exchangeSinkHandle);
+                    });
+                }
+                if (taskSource.isFinished()) {
+                    sinkExchange.ifPresent(Exchange::noMoreSinks);
+                }
+            }
+
+            if (queuedPartitions.isEmpty()) {
+                break;
+            }
+
+            int partition = queuedPartitions.peek();
+            TaskDescriptor taskDescriptor = requireNonNull(partitionToTaskDescriptorMap.get(partition), () -> "task descriptor missing for partition: %s" + partition);
+
+            if (acquireNodeFuture == null) {
+                acquireNodeFuture = nodeAllocator.acquire(taskDescriptor.getNodeRequirements());
+            }
+            if (!acquireNodeFuture.isDone()) {
+                blocked = asVoid(acquireNodeFuture);
+                return;
+            }
+            InternalNode node = getFutureValue(acquireNodeFuture);
+            acquireNodeFuture = null;
+
+            queuedPartitions.poll();
+
+            Multimap<PlanNodeId, Split> tableScanSplits = taskDescriptor.getSplits();
+            Multimap<PlanNodeId, Split> remoteSplits = createRemoteSplits(taskDescriptor.getExchangeSourceHandles());
+
+            Multimap<PlanNodeId, Split> taskSplits = ImmutableListMultimap.<PlanNodeId, Split>builder()
+                    .putAll(tableScanSplits)
+                    .putAll(remoteSplits)
+                    .build();
+
+            int attemptId = getNextAttemptIdForPartition(partition);
+
+            OutputBuffers outputBuffers;
+            Optional<ExchangeSinkInstanceHandle> exchangeSinkInstanceHandle;
+            if (sinkExchange.isPresent()) {
+                ExchangeSinkHandle sinkHandle = partitionToExchangeSinkHandleMap.get(partition);
+                exchangeSinkInstanceHandle = Optional.of(sinkExchange.get().instantiateSink(sinkHandle, attemptId));
+                outputBuffers = createExternalExchangeOutputBuffers(exchangeSinkInstanceHandle.get());
+            }
+            else {
+                // stage will be consumed by the coordinator using direct exchange
+                outputBuffers = createInitialEmptyOutputBuffers(PARTITIONED)
+                        .withBuffer(new OutputBuffers.OutputBufferId(0), 0)
+                        .withNoMoreBufferIds();
+                exchangeSinkInstanceHandle = Optional.empty();
+            }
+
+            Set<PlanNodeId> allSourceNodeIds = new HashSet<>(stageExecution.getFragment().getPartitionedSources());
```

Comment:
this class is operating both on plan-nodes and cluster-nodes. Please be explicit about that in variable names.
Lets use `*planNode*` for plan-nodes and plain `*Node*` for cluster nodes.